src/audio/rtnr/rtnr.c

```diff
@@ -0,0 +1,549 @@
+{
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_SWITCH:
+		comp_dbg(dev, "rtnr_cmd_set_value(), SOF_CTRL_CMD_SWITCH, cdata->comp_id = %u",
+			 cdata->comp_id);
+		break;
+
+	default:
+		comp_err(dev, "rtnr_cmd_set_value() error: invalid cdata->cmd");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+/* used to pass standard and bespoke commands (with data) to component */
+static int rtnr_cmd(struct comp_dev *dev, int cmd, void *data,
+		   int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+	comp_info(dev, "rtnr_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = rtnr_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = rtnr_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	case COMP_CMD_SET_VALUE:
+		ret = rtnr_cmd_set_value(dev, cdata);
+		break;
+	case COMP_CMD_GET_VALUE:
+		ret = rtnr_cmd_get_value(dev, cdata);
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd() error: invalid command");
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+static int rtnr_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_info(dev, "rtnr_trigger() cmd: %d", cmd);
+
+	return comp_set_state(dev, cmd);
+}
+
+/* copy and process stream data from source to sink buffers */
+static int rtnr_copy(struct comp_dev *dev)
+{
+	struct comp_buffer *sink;
+	struct comp_buffer *source;
+	int frames;
+	int num_blocks;
+	int sink_bytes;
+	int source_bytes;
+	const struct audio_stream *sources_stream[RTNR_MAX_SOURCES] = { NULL };
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "rtnr_copy()");
+
+	source = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+
+	/* put empty data into output queue*/
+	int nch = source->stream.channels;
+	/* RTKMA uses 32 bits internal storage */
+	int copy_size = BLOCKLENGTH * nch * sizeof(int32_t);
+
+	RTKMA_API_First_Copy((void *)&cd->rtk_agl, copy_size);
+
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	frames = audio_stream_avail_frames(&source->stream, &sink->stream);
+
+	/* Process integer multiple of RTNR internal block length */
+	num_blocks = frames / RTNR_BLK_LENGTH;
+	frames = num_blocks * RTNR_BLK_LENGTH;
+
+	comp_dbg(dev, "rtnr_copy() source->id: %d, frames = %d", source->id, frames);
+
+	if (frames) {
+		source_bytes = frames * audio_stream_frame_bytes(&source->stream);
+		sink_bytes = frames * audio_stream_frame_bytes(&sink->stream);
+
+		buffer_invalidate(source, source_bytes);
+
+		/* Run processing function */
+
+		/*
+		 * Processing function uses an array of pointers to source streams
+		 * as parameter.
+		 */
+		sources_stream[0] = &source->stream;
+		cd->rtnr_func(dev, sources_stream, &sink->stream, frames);
+
+		buffer_writeback(sink, sink_bytes);
+
+		/*real process function of rtk ma*/
+		/* RTKMA uses 32 bits internal storage */
+		copy_size = BLOCKLENGTH * MicNum * sizeof(int32_t);
+		RTKMA_API_Process((void *)&cd->rtk_agl, copy_size, 0);
```

Comment:
can this fail?

Reply:
Also im still confused on what this does, does it produce samples in the sink at all?

Reply:
RTKMA_API_Process is the main processing function. The actual function to consume data from source or output data to sink are RTKMA_API_(S16|S24|S32)_Default

Reply:
RTKMA_API_Process is now called before buffer writeback, thanks.